internal/era2/builder2.go

```diff
@@ -0,0 +1,321 @@
+	ProofRoots   proofvar = proofvar(TypeProofHistoricalRoots)
+	ProofCapella proofvar = proofvar(TypeProofHistoricalSummariesCapella)
+	ProofDeneb   proofvar = proofvar(TypeProofHistoricalSummariesDeneb)
+)
+
+// Proof bundles variant + compressed bytes.
+type Proof struct {
+	Variant proofvar
+	Data    []byte
+}
+
+type Builder struct {
+	w   *e2store.Writer
+	buf *bytes.Buffer
+	sn  *snappy.Writer
+
+	// buffered entries per type:
+	headersRLP  [][]byte
+	bodiesRLP   [][]byte
+	receiptsRLP [][]byte
+	proofsRLP   [][]byte
+	tds         [][]byte
+	tdsint      []*big.Int
+	hashes      []common.Hash
+
+	headeroffsets  []uint64
+	bodyoffsets    []uint64
+	receiptoffsets []uint64
+	proofoffsets   []uint64
+	tdoff          []uint64
+	startTd        *big.Int
+
+	prooftype proofvar
+
+	startNum     *uint64
+	writtenBytes uint64
+}
+
+func NewBuilder(w io.Writer) *Builder {
+	buf := bytes.NewBuffer(nil)
+	return &Builder{
+		w:   e2store.NewWriter(w),
+		buf: buf,
+		sn:  snappy.NewBufferedWriter(buf),
+	}
+}
+
+func (b *Builder) Add(header types.Header, body types.Body, receipts types.Receipts, blockhash common.Hash, blocknum uint64, td *big.Int, proof *Proof) error {
+	if len(b.headersRLP) >= MaxEraESize {
+		return fmt.Errorf("exceeds MaxEraESize %d", MaxEraESize)
+	}
+
+	var pv proofvar = ProofNone
+	var pData []byte
+	if proof != nil {
+		pv = proof.Variant
+		pData = proof.Data
+		if pv == ProofNone || len(pData) == 0 {
+			return fmt.Errorf("invalid proof: variant=%d len=%d", pv, len(pData))
+		}
+	}
+
+	if len(b.headersRLP) == 0 {
+		b.prooftype = pv
+	} else if pv != b.prooftype {
+		return fmt.Errorf("cannot mix proof variants: have %d want %d", b.prooftype, pv)
+	}
+
+	hdr, err := rlp.EncodeToBytes(&header)
+	if err != nil {
+		return fmt.Errorf("error encoding block header: %w", err)
+	}
+	bod, err := rlp.EncodeToBytes(&body)
+	if err != nil {
+		return fmt.Errorf("error encoding block header: %w", err)
+	}
+	rct, err := rlp.EncodeToBytes(receipts)
+	if err != nil {
+		return fmt.Errorf("error encoding block header: %w", err)
+	}
+
+	b.headersRLP = append(b.headersRLP, hdr)
+	b.bodiesRLP = append(b.bodiesRLP, bod)
+	b.receiptsRLP = append(b.receiptsRLP, rct)
+	b.tds = append(b.tds, uint256LE(td))
+	b.tdsint = append(b.tdsint, new(big.Int).Set(td))
+	b.hashes = append(b.hashes, blockhash)
+
+	if b.prooftype != ProofNone {
+		b.proofsRLP = append(b.proofsRLP, proof.Data)
+	}
+
+	if b.startNum == nil {
+		sn := blocknum
+		b.startNum = &sn
+		if n, err := b.w.Write(TypeVersion, nil); err != nil {
+			return fmt.Errorf("error writing version entry: %w", err)
+		} else {
+			b.writtenBytes += uint64(n)
+		}
+	}
+	return nil
+}
+
+func (b *Builder) Finalize() (common.Hash, error) {
+	if b.startNum == nil {
+		return common.Hash{}, errors.New("no blocks added, cannot finalize")
+	}
+	var err error
+	if err := b.flushKind(TypeCompressedHeader, b.headersRLP, true, &b.headeroffsets); err != nil {
```

Comment:
It's a bit weird that `flushKind` is a method on `Builder` but 2/4 parameters are just variables inside the `Builder`. Probably better to refactor it so it also takes the `io.Writer` from `Builder` and is then just a separate method. Would also consider hoisting the logic back up here - a loop of writes isn't that complicated and might be easier to read than 2-3 levels of indirection to write the values.